azure_scripts/lotka_volterra_francois.py

```diff
@@ -0,0 +1,252 @@
+# create simulations
+@jax.jit
+def get_batch(key, batch_size=args.batch_size):
+    model = condition(lotka_volterra, {'z': jnp.array([30.,1.])})
+    (log_probs, samples), scores = get_samples_and_scores(model, key, batch_size=batch_size)
+    return samples['theta'], samples['y'].reshape([-1,20]), scores
+
+# create model
+bijector_layers = [args.bijector_layers_size] * args.bijector_layers_shape
+
+bijector = partial(
+  AffineSigmoidCoupling, 
+  layers=bijector_layers, 
+  n_components=args.n_components, 
+  activation=jnp.sin
+)
+
+NF = partial(
+  ConditionalRealNVP, 
+  n_layers=args.nf_layers, 
+  bijector_fn=bijector
+)
+
+class SmoothNPE(hk.Module):
+    def __call__(self, y):
+        net = lotka_volterra_y_bijector(y)
+        net = jnp.sin(hk.Linear(256, name='comp1')(net))
+        net = jnp.sin(hk.Linear(256, name='comp2')(net))
+        net = jnp.sin(hk.Linear(256, name='comp3')(net))
+        net = hk.Linear(args.n_components, name='comp4')(net)
+        nvp = NF(args.dimension)(net)
+        return tfd.TransformedDistribution(nvp,
+                                           tfb.Chain([tfb.Invert(lotka_volterra_theta_bijector),
+                                                      tfb.Scale(10.),
+                                                      tfb.Shift(-0.5)]))
+
+nvp_nd = hk.without_apply_rng(hk.transform(lambda theta,y : SmoothNPE()(y).log_prob(theta).squeeze()))
+
+# init parameters
+params_nd = nvp_nd.init(next(rng_seq), jnp.ones([1,4]), jnp.ones([1,20]))
+
+# init optimizer
+scheduler = optax.piecewise_constant_schedule(init_value=args.initial_learning_rate, 
+                                              boundaries_and_scales={5000: 0.1,
+                                                                     10000: 0.5,
+                                                                     15000: 0.1})
+optimizer = optax.adam(learning_rate=scheduler)
+opt_state = optimizer.init(params_nd)
+
+# define loss function and model update
+def loss_fn(params, mu, batch, score, score_weight):
+  log_prob, out = jax.vmap(
+    jax.value_and_grad(lambda theta, x: nvp_nd.apply(params, theta.reshape([1, 4]), x.reshape([1, -1])).squeeze())
+  )(mu, batch)
+  return -jnp.mean(log_prob) + score_weight * jnp.mean(jnp.sum((out - score)**2, axis=1))
+
+
+@jax.jit
+def update(params, opt_state, mu, batch, score, score_weight=args.score_weight):
+  """Single SGD update step."""
+  loss, grads = jax.value_and_grad(loss_fn)(params, mu, batch, score, score_weight)
+  updates, new_opt_state = optimizer.update(grads, opt_state)
+  new_params = optax.apply_updates(params, updates)
+
+  return loss, new_params, new_opt_state
+
+# train
+batch_loss = []
+for step in range(args.n_steps):
+    mu, batch, score = get_batch(next(rng_seq))
+    l, params_nd, opt_state = update(params_nd, opt_state, mu, batch, score)
+    if (step % 100) == 0 :
+        print(f'Iter {step} / {args.n_steps}:', np.mean(batch_loss[50:]), scheduler(opt_state[1].count))
```

Comment:
@EiffL is there a reason you compute the mean over all the steps starting at 50 and not the mean over the latest 50 or 100 ?

Reply:
oh right ^^ that's not what I meant to do. mean over the latest 50 is the correct thing